kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -167,6 +167,14 @@ public Transaction build(Engine engine) {
     /* ----- 2: Update the PROTOCOL based on the table properties or schema ----- */
     // This is the only place we update the protocol action; takes care of any dependent features
     // Ex: We enable feature `icebergCompatV2` plus dependent features `columnMapping`
+    Optional<Tuple2<Protocol, Metadata>> newProtocolAndMetadata =
```

Comment:
I'm not sure how this works considering below we still use `snapshotProtocol` for `autoUpgradeProtocolBasedOnMetadata`

Reply:
I think your tests pass because we only use this code. Please add an E2E test case where we are manually enabling a feature AND enabling one via metadata (which currently would not succeed -- verify this)

Reply:
But can I actually suggest a different approach? Can we in step 1, do this filtering out, and build a list called `manuallyEnabledTableFeatures` and pass this as an argument into `autoUpgradeProtocolBasedOnMetadata`?

We can then also include `needsDomainMetadataSupport` and `needsClusteringTableFeature` all into that 1 argument.

And then --> let's make sure we also enable any _dependent features_ for this list of manually enabled table features (something this currently doesn't do I think) in `autoUpgradeProtocolBasedOnMetadata`.

Reply:
And let's also make sure to have a test case where we manually enable a feature with dependent features :) so like manually enable `icebergCompatV2`

Reply:
Let me know if this approach makes sense -- I'm open to objections if you don't agree

Reply:
I think we need to fix the input protocol to `autoUpgradeProtocolBasedOnMetadata`. Regarding the approach @allisonport-db , I think it makes sense to me.